batcher/sweep_batch.go

```diff
@@ -0,0 +1,1227 @@
+		if err != nil {
+			return fee, err
+		}
+
+		// Append this sweep to an array of sweeps. This is needed to
+		// keep the order of sweeps stored, as iterating the sweeps map
+		// does not guarantee same order.
+		sweeps = append(sweeps, sweep)
+
+		// Create and store the previous outpoint for this sweep.
+		prevOuts = append(prevOuts, &wire.TxOut{
+			Value:    int64(sweep.Value),
+			PkScript: sweep.Htlc.PkScript,
+		})
+
+		key, err := btcec.ParsePubKey(sweep.Key[:])
+		if err != nil {
+			return fee, err
+		}
+
+		// Create and store the sign descriptor for this sweep.
+		signDesc := lndclient.SignDescriptor{
+			WitnessScript: sweep.Htlc.SuccessScript(),
+			Output:        prevOuts[len(prevOuts)-1],
+			HashType:      sweep.Htlc.SigHash(),
+			InputIndex:    inputCounter,
+			KeyDesc: keychain.KeyDescriptor{
+				PubKey: key,
+			},
+		}
+
+		inputCounter++
+
+		if sweep.Htlc.Version == swap.HtlcV3 {
+			signDesc.SignMethod = input.TaprootScriptSpendSignMethod
+		}
+
+		signDescs = append(signDescs, &signDesc)
+	}
+
+	// Generate a wallet address for the batch transaction's output.
+	address, err := b.wallet.NextAddr(
+		ctx, "", walletrpc.AddressType_TAPROOT_PUBKEY, false,
+	)
+	if err != nil {
+		return fee, err
+	}
+
+	batchPkScript, err := txscript.PayToAddrScript(address)
+	if err != nil {
+		return fee, err
+	}
+
+	weightEstimate.AddP2TROutput()
+
+	totalWeight := int64(weightEstimate.Weight())
+
+	fee = b.RbfCache.FeeRate.FeeForWeight(totalWeight)
+
+	// Add the batch transaction output, which excludes the fees paid to
+	// miners.
+	batchTx.AddTxOut(&wire.TxOut{
+		PkScript: batchPkScript,
+		Value:    int64(batchAmt - fee),
+	})
+
+	// Collect the signatures for our inputs.
+	rawSigs, err := b.signerClient.SignOutputRaw(
+		ctx, batchTx, signDescs, prevOuts,
+	)
+	if err != nil {
+		return fee, err
+	}
+
+	for i, sweep := range sweeps {
+		// Generate the success witness for the sweep.
+		witness, err := sweep.Htlc.GenSuccessWitness(
+			rawSigs[i], sweep.Preimage,
+		)
+		if err != nil {
+			return fee, err
+		}
+
+		// Add the success witness to our batch transaction's inputs.
+		batchTx.TxIn[i].Witness = witness
+	}
+
+	b.log.Debugf("attempting to publish tx with feerate=%v, totalfee=%v",
+		b.RbfCache.FeeRate, fee)
+	err = b.wallet.PublishTransaction(
+		ctx, batchTx, labels.LoopOutBatchSweepSuccess(b.ID),
+	)
+	if err != nil {
+		return fee, err
+	}
+
+	// Store the batch transaction's txid and pkScript, for monitoring
+	// purposes.
+	txHash := batchTx.TxHash()
+	b.BatchTxid = &txHash
+	b.BatchPkScript = batchPkScript
+
+	return fee, nil
+}
+
+// publishBatchCoop attempts to construct and publish a batch transaction that
+// collects all the required signatures interactively from the server. This
+// helps with collecting the funds immediately without revealing any information
+// related to the HTLC script.
+func (b *Batch) publishBatchCoop(ctx context.Context) (btcutil.Amount,
```

Comment:
Could we break this function up a bit? Essentially  the signing part could be a separate function for readability.